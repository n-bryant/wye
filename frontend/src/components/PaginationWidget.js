import React from "react";
import PropTypes from "prop-types";

import { withStyles } from "@material-ui/core/styles";
import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../lib/classNamePrefix";

import Icon from "@mdi/react";
import {
  mdiChevronLeft,
  mdiChevronRight,
  mdiPageFirst,
  mdiPageLast
} from "@mdi/js";

import Button from "@material-ui/core/Button";
import Typography from "@material-ui/core/Typography";
import styles from "./PaginationWidget.styles";

/**
 * renders a pagination widget with prev/next buttons and an first/last page buttons
 */
export const PaginationWidget = props => {
  const classnames = PaginationWidget.classnames(props);
  const { currentPage, setCurrentPage, totalPages } = props;

  // paddle disabled logic
  const prevPaddleDisabled = currentPage === 1;
  const nextPaddleDisabled = currentPage === totalPages;

  return (
    <div className={classnames.root()}>
      {totalPages > 2 && (
        <Button
          className={classnames.element("paddle")}
          disabled={prevPaddleDisabled}
          onClick={() => setCurrentPage(1)}
        >
          <Icon
            className={classnames.element("paddleIcon", {
              disabled: prevPaddleDisabled
            })}
            path={mdiPageFirst}
          />
        </Button>
      )}
      <Button
        className={classnames.element("paddle")}
        disabled={prevPaddleDisabled}
        onClick={() => setCurrentPage(currentPage - 1)}
      >
        <Icon
          className={classnames.element("paddleIcon", {
            disabled: prevPaddleDisabled
          })}
          path={mdiChevronLeft}
        />
      </Button>
      <Typography variant="body2">{`${currentPage} of ${totalPages}`}</Typography>
      <Button
        className={classnames.element("paddle")}
        disabled={nextPaddleDisabled}
        onClick={() => setCurrentPage(currentPage + 1)}
      >
        <Icon
          className={classnames.element("paddleIcon", {
            disabled: nextPaddleDisabled
          })}
          path={mdiChevronRight}
        />
      </Button>
      {totalPages > 2 && (
        <Button
          className={classnames.element("paddle")}
          disabled={nextPaddleDisabled}
          onClick={() => setCurrentPage(totalPages)}
        >
          <Icon
            className={classnames.element("paddleIcon", {
              disabled: nextPaddleDisabled
            })}
            path={mdiPageLast}
          />
        </Button>
      )}
    </div>
  );
};
PaginationWidget.classnames = createClassNameHelper(
  `${JSS_CLASS_NAME_PREFIX}PaginationWidget`
);
PaginationWidget.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    paddlesWrapper: PropTypes.string,
    paddle: PropTypes.string,
    paddleIcon: PropTypes.string,
    paddleIconDisabled: PropTypes.string,
    pageIndicator: PropTypes.string,
    pagingValuesContainer: PropTypes.string,
    goToPageInput: PropTypes.string
  }),
  // the current page number
  currentPage: PropTypes.number,
  // handler to set the current page
  setCurrentPage: PropTypes.func,
  // the count of total pages
  totalPages: PropTypes.number
};
PaginationWidget.defaultProps = {
  classes: {},
  currentPage: 1,
  setCurrentPage: () => {}
};

export default withStyles(styles)(PaginationWidget);
