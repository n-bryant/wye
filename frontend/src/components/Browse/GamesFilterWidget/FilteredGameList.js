import React from "react";
import PropTypes from "prop-types";

import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../../../lib/classNamePrefix";
import { withStyles } from "@material-ui/core/styles";

import { useRouter } from "next/router";

import Box from "@material-ui/core/Box";
import Container from "@material-ui/core/Container";
import Grid from "@material-ui/core/Grid";
import TextField from "@material-ui/core/TextField";
import Autocomplete from "@material-ui/lab/Autocomplete";

import {
  getOffsetStart,
  getOffsetEnd,
  getTotalPages
} from "../../../../lib/pagingUtilities";
import BrowseCard from "../BrowseCard";
import PaginationWidget from "../../PaginationWidget";
import styles from "./FilteredGameList.styles";

/**
 * renders a list BrowseCards to represent the provided list of filtered game items
 */
export const FilteredGameList = props => {
  const classnames = FilteredGameList.classnames(props);
  const { classes, items, userDetails } = props;
  const router = useRouter();

  // pagination values
  const perPage = 20;
  const [currentPage, setCurrentPage] = React.useState(1);
  const totalPages = getTotalPages(items.length, perPage);
  const pageStart = getOffsetStart(currentPage, items.length, perPage);
  const pageEnd = getOffsetEnd(currentPage, perPage);

  return (
    <Container
      className={classnames.element("root")}
      maxWidth="lg"
      disableGutters={true}
    >
      <Box my={4}>
        <Autocomplete
          classes={{
            paper: classes.paper,
            option: classes.option
          }}
          id="filtered-game-list-search"
          onChange={(_e, value) => {
            router.push(`/game/${value.appid}`);
          }}
          options={items.map(item => item.node.game)}
          getOptionLabel={item => item.name}
          style={{ width: 300 }}
          renderInput={params => (
            <TextField {...params} label="Search" fullWidth />
          )}
        />
      </Box>
      <Box my={4}>
        {totalPages > 1 && (
          <Box my={4}>
            <PaginationWidget
              currentPage={currentPage}
              setCurrentPage={setCurrentPage}
              totalPages={totalPages}
            />
          </Box>
        )}
        <Grid container justify="center" spacing={6}>
          {items.slice(pageStart, pageEnd).map(item => (
            <Grid
              key={item.node.game.appid}
              item
              xs={12}
              md={6}
              className={classnames.element("item")}
            >
              <BrowseCard
                data={item.node.game}
                playtimeAndOwnershipData={{
                  ownedBy: item.node.ownedBy,
                  recentlyPlayedBy: item.node.recentlyPlayedBy,
                  playtime: item.node.playtime
                }}
                userDetails={userDetails}
                withPrice={true}
                variant="horizontal"
              />
            </Grid>
          ))}
        </Grid>
        {totalPages > 1 && (
          <Box my={4}>
            <PaginationWidget
              currentPage={currentPage}
              setCurrentPage={setCurrentPage}
              totalPages={totalPages}
            />
          </Box>
        )}
      </Box>
    </Container>
  );
};
FilteredGameList.classnames = createClassNameHelper(
  `${JSS_CLASS_NAME_PREFIX}FilteredGameList`
);
FilteredGameList.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    item: PropTypes.string,
    paper: PropTypes.string,
    option: PropTypes.string
  }),
  // the items to render in the list
  items: PropTypes.array,
  // details for users associated with the items
  userDetails: PropTypes.array
};
FilteredGameList.defaultProps = {
  classes: {}
};

// apply styles
export const StyledFilteredGameList = withStyles(styles)(FilteredGameList);

export default StyledFilteredGameList;
