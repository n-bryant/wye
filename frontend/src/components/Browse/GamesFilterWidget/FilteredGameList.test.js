import PaginationWidget from "../../PaginationWidget";
import { FilteredGameList } from "./FilteredGameList";

describe("FilteredGameList", () => {
  const props = {
    classes: {
      root: "root",
      item: "item"
    },
    items: [
      {
        node: {
          game: {
            appid: "1"
          }
        }
      }
    ]
  };

  it("should render successfully", () => {
    const wrapper = shallow(<FilteredGameList {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should render pagination widgets if the number of pages is greater than 1", () => {
    let mockedItems = [];
    for (let i = 0; i < 22; i++) {
      mockedItems.push({
        node: {
          game: {
            appid: `${i}`
          }
        }
      });
    }
    const propsWithPages = {
      ...props,
      items: mockedItems
    };
    const wrapper = shallow(<FilteredGameList {...propsWithPages} />);
    expect(wrapper.find(PaginationWidget).length).toBe(2);
  });
});
