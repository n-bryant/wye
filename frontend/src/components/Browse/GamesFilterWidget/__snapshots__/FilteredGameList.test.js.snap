// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilteredGameList should render successfully 1`] = `
<WithStyles(ForwardRef(Container))
  className="wye-FilteredGameList__root root"
  disableGutters={true}
  maxWidth="lg"
>
  <Styled(MuiBox)
    my={4}
  >
    <WithStyles(ForwardRef(Grid))
      container={true}
      justify="center"
      spacing={6}
    />
  </Styled(MuiBox)>
</WithStyles(ForwardRef(Container))>
`;
